SYCL/ESIMD/api/functional/ctors/ctor_move.cpp

```diff
@@ -0,0 +1,233 @@
+        const auto ptr_out = result.data();
+        const auto ptr_flags = flags_storage.data();
+
+        cgh.parallel_for<KernelName>(
+            range, [=](sycl::id<1> id) SYCL_ESIMD_KERNEL {
+              const auto work_item_index = id[0];
+              // Access a separate memory areas from each of the work items
+              const DataT *const in = ptr_in + work_item_index;
+              const auto out = ptr_out + work_item_index;
+              const auto was_moved_flag = ptr_flags + work_item_index;
+              *was_moved_flag = run_check(in, out);
+            });
+      });
+      queue.wait_and_throw();
+
+      // Oversafe: verify the proper signature was called for every check
+      was_moved = std::all_of(flags_storage.cbegin(), flags_storage.cend(),
+                              [](int flag) { return flag; });
+    } else {
+      assert((input.size() == NumElems) &&
+             "Unexpected size of the input vector");
+
+      shared_vector<int> flags_storage(1, flags_allocator);
+
+      queue.submit([&](sycl::handler &cgh) {
+        const DataT *const in = input.data();
+        const auto out = result.data();
+        const auto was_moved_flag = flags_storage.data();
+
+        cgh.single_task<KernelName>(
+            [=]() SYCL_ESIMD_KERNEL { *was_moved_flag = run_check(in, out); });
+      });
+      queue.wait_and_throw();
+
+      was_moved = flags_storage[0];
+    }
+
+    if (!was_moved) {
+      passed = false;
+
+      // TODO: Make ITestDescription architecture more flexible
+      std::string log_msg = "Failed for simd<";
+      log_msg += data_type + ", " + std::to_string(NumElems) + ">";
+      log_msg += ", with context: " + TestCaseT::get_description();
+      log_msg += ". A copy constructor instead of a move constructor was used.";
+
+      log::note(log_msg);
+    } else {
+      for (size_t i = 0; i < reference.size(); ++i) {
+        const auto &retrieved = result[i];
+        const auto &expected = reference[i];
+
+        if (!are_bitwise_equal(retrieved, expected)) {
+          passed = false;
+
+          log::fail(ctors::TestDescription<DataT, NumElems, TestCaseT>(
+              i, retrieved, expected, data_type));
+        }
+      }
+    }
+
+    return passed;
+  }
+
+private:
+  // The core check logic.
+  // Uses USM pointers for input data and to store the data from the new simd
+  // instance, so that we could check it later
+  // Returns the flag that should be true only if the move constructor was
+  // actually called, to differentiate with the copy constructor calls
+  static bool run_check(const DataT *const in, DataT *const out) {
+    bool was_moved = false;
+
+    // Prepare the source simd to move
+    simd<DataT, NumElems> source;
+    source.copy_from(in);
+
+    // Action to run over the simd move constructor result
+    const auto action = [&](const simd<DataT, NumElems> &instance) {
+      was_moved = instance.get_test_proxy().was_move_destination();
```

Comment:
It is odd, but I cannot find the definition of 'was_move_destination()'. Would you please give me a link to it?

Reply:
Sure, sorry for any possible inconvenience.
Both the `test_proxy` class and the `test_proxy::was_move_destination()` are from the intel/llvm#5132.
Here is a link to the exact [line](https://github.com/intel/llvm/blob/f0465df6ee94da25704d82902649e74b04f85dd8/sycl/include/sycl/ext/intel/experimental/esimd/detail/test_proxy.hpp#L99).

Reply:
Thank you.